feat(notebooks): add variables and Blob Storage API documentation (#24904)
newrelic/docs-website: src/content/docs/query-your-data/explore-query-data/notebooks/blob-storage-api-for-notebooks.mdx

---
title: 'Using the Blob Storage API for notebooks'
tags:
  - Query your data
  - Explore and query data
  - Notebooks
  - API
metaDescription: 'Use the New Relic Blob Storage API and NerdGraph to create, read, update, delete, and version notebooks programmatically.'
freshnessValidatedDate: never
---

> <Callout title="Public preview">

Reviewer: Could you please check the callout as well please

Author: It's required

  The <DNT>Blob Storage API</DNT> for notebooks is currently in public preview. This feature is provided pursuant to our [pre-release policies](/docs/licenses/license-information/referenced-policies/new-relic-pre-release-policy).
</Callout>

> The <DNT>New Relic</DNT> Notebooks API lets you create, read, update, and delete notebooks programmatically, including their full block content (<DNT>NRQL</DNT> queries, text). Notebooks are stored as versioned blobs, which means every save produces a new immutable revision you can retrieve later.

Reviewer: Notebooks is inconsistent, in few places it is Capital case and other places it is mentioned notbooks

> The <DNT>New Relic</DNT> Notebooks API lets you create, read, update, and delete notebooks programmatically, including their full block content (<DNT>NRQL</DNT> queries, text). Notebooks are stored as versioned blobs, which means every save produces a new immutable revision you can retrieve later.

Reviewer: ```suggestion
The New Relic <DNT>Notebooks</DNT> API lets you create, read, update, and delete notebooks programmatically, including their full block content (<DNT>NRQL</DNT> queries, text). Notebooks are stored as versioned blobs, which means every save produces a new immutable revision you can retrieve later.
```

Use this API to:

* Automate notebook creation from incident templates, runbooks, or CI/CD pipelines
* Sync notebook content from version control or external authoring tools
* Build integrations that programmatically populate notebooks during investigations

<Callout variant="important">

>   **Notebooks use multiple APIs**

Reviewer: ```suggestion
  **<DNT>Notebooks</DNT> use multiple APIs**
```

>   The Notebooks API surface is split between two systems:

Reviewer: ```suggestion
  The <DNT>Notebooks</DNT> API surface is split between two systems:
```

  * **<DNT>Blob Storage API</DNT>** handles notebook content (blocks, version history)
  * **<DNT>NerdGraph</DNT>** handles entity-level operations (list, rename, tags, organization metadata)

  This separation is by design. The <DNT>Blob Storage API</DNT> is optimized for file content transfer and versioning; <DNT>NerdGraph</DNT> is optimized for structured entity queries and mutations.
</Callout>

## Prerequisites [#prerequisites]

* A [<DNT>New Relic</DNT> account](https://newrelic.com/signup) with a User API key
* Your <DNT>New Relic</DNT> Organization ID
* Appropriate [permissions](/docs/accounts/accounts-billing/new-relic-one-user-management/user-permissions/) to manage notebooks

## Authentication [#authentication]

> All Notebook API requests require authentication using a <DNT>New Relic</DNT> User API key.

Reviewer: ```suggestion
All <DNT>Notebooks</DNT> API requests require authentication using a <DNT>New Relic</DNT> User API key.
```

**Generate an API key:**

1. Navigate to [one.newrelic.com](https://one.newrelic.com)

> 2. Click on your name in the bottom-left corner

Reviewer: ```suggestion
2. Click on your name in the top-right corner
```

3. Select <DNT>**API Keys**</DNT>
4. Create a <DNT>**User**</DNT> key (not Browser or License key)

**Include in request headers:**

```shell
Api-Key: NRAK-YOUR-USER-API-KEY
```

<Callout variant="tip">
  The <DNT>Blob Storage API</DNT> also supports login context, so when calling the API from a UI authenticated as a <DNT>New Relic</DNT> user, the `Api-Key` header is not required.
</Callout>

## Base endpoint [#base-endpoint]

```plaintext
https://blob-api.service.newrelic.com/v1/e
```

For EU region accounts, use:

```plaintext
https://blob-api.service.eu.newrelic.com/v1/e
```

## Notebook content operations [#content-operations]

<CollapserGroup>

<Collapser
  id="create-notebook"
  title="Create notebook">

Creates a new notebook entity with initial blob content.

### Endpoint [#create-endpoint]

```plaintext
POST /v1/e/organizations/{orgId}/Notebooks
```

### Request parameters [#create-params]

<table>
  <thead>
    <tr>
      <th>Parameter</th>
      <th>Location</th>
      <th>Data type</th>

>       <th>Is it required?</th>

Reviewer: ```suggestion

```

      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>`orgId`</td>
      <td>Path</td>
      <td>String</td>

>       <td>Yes</td>

Reviewer: ```suggestion

```

>       <td>Your <DNT>New Relic</DNT> organization ID.</td>

Reviewer: ```suggestion
      <td>Required. Your <DNT>New Relic</DNT> organization ID.</td>
```

    </tr>
    <tr>
      <td>`Api-Key`</td>
      <td>Header</td>
      <td>String</td>

>       <td>Yes</td>

Reviewer: ```suggestion

```

>       <td>Your User API key.</td>

Reviewer: ```suggestion
      <td>Required. Your User API key.</td>
```

    </tr>
    <tr>
      <td>`Content-Type`</td>
      <td>Header</td>
      <td>String</td>

>       <td>Yes</td>

Reviewer: ```suggestion

```

>       <td>Must be `application/json`.</td>

Reviewer: ```suggestion
      <td>Required. Must be `application/json`.</td>
```

    </tr>
    <tr>
      <td>`NewRelic-Entity`</td>
      <td>Header</td>
      <td>JSON string</td>

>       <td>Yes</td>

Reviewer: ```suggestion

```

>       <td>JSON object with notebook entity metadata (see below).</td>

Reviewer: ```suggestion
      <td>Required. JSON object with notebook entity metadata (see below).</td>
```

    </tr>
    <tr>
      <td>Request body</td>
      <td>Body</td>
      <td>JSON</td>

>       <td>Yes</td>

Reviewer: ```suggestion

```

>       <td>Notebook content in JSON format (version + blocks).</td>

Reviewer: ```suggestion
      <td>Required. Notebook content in JSON format (version + blocks).</td>
```

    </tr>
  </tbody>
</table>

### `NewRelic-Entity` header format [#create-entity-header]

<table>
  <thead>
    <tr>
      <th>Field</th>
      <th>Data type</th>

>       <th>Is it required?</th>

Reviewer: ```suggestion

```

      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>`name`</td>
      <td>String</td>

>       <td>Yes</td>

Reviewer: ```suggestion

```

>       <td>The notebook name. Must be unique within the organization scope.</td>

Reviewer: ```suggestion
      <td>Required. The notebook name. Must be unique within the organization scope.</td>
```

```plaintext
DELETE /v1/e/organizations/{orgId}/Notebooks/{entityGuid}
```

### Request parameters [#delete-params]

<table>
  <thead>
    <tr>
      <th>Parameter</th>
      <th>Location</th>
      <th>Data type</th>
      <th>Is it required?</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>`orgId`</td>
      <td>Path</td>
      <td>String</td>
      <td>Yes</td>
      <td>Your <DNT>New Relic</DNT> organization ID.</td>
    </tr>
    <tr>
      <td>`entityGuid`</td>
      <td>Path</td>
      <td>String</td>
      <td>Yes</td>
      <td>The notebook entity GUID to delete.</td>
    </tr>
    <tr>
      <td>`Api-Key`</td>
      <td>Header</td>
      <td>String</td>
      <td>Yes</td>
      <td>Your User API key.</td>
    </tr>
  </tbody>
</table>

### Sample request [#delete-sample-request]

```bash
curl -X DELETE \
  https://blob-api.service.newrelic.com/v1/e/organizations/YOUR_ORG_ID/Notebooks/YOUR_ENTITY_GUID \
  -H 'Api-Key: NRAK-YOUR-API-KEY'
```

### Sample response [#delete-sample-response]

Returns HTTP 204 No Content on successful deletion.

</Collapser>

<Collapser
  id="retrieve-versions"
  title="Retrieve previous versions">

> Notebook versions are retained for **1 day**. To recover a previous revision, first list recent versions via <DNT>NRQL</DNT>, then fetch the specific blob.

Reviewer: ```suggestion
<DNT>Notebook</DNT> versions are retained for **1 day**. To recover a previous revision, first list recent versions via <DNT>NRQL</DNT>, then fetch the specific blob.
```

    notebookEntity: { name: "<new name>" }
  ) {
    entity {
      name
    }
  }
}
```

### Update notebook tags [#update-tags]

<Callout variant="important">
  Tag updates are a **replace** operation. You must include the complete set of tags, even ones that aren't changing — any tag omitted from the mutation will be removed.
</Callout>

```graphql
mutation updateNotebookTags {
  entityManagementUpdateNotebook(
    id: "<entity guid>"
    notebookEntity: {
      tags: [
        { key: "<key>", values: "<value>" }
        { key: "<key>", values: "<value>" }
      ]
    }
  ) {
    entity {
      name
      tags {
        key
        values
      }
    }
  }
}
```

### Retrieve your organization ID [#get-org-id]

You'll need your organization ID for all <DNT>Blob Storage API</DNT> calls:

```graphql
query getOrgId {
  actor {
    organization {
      id
    }
  }
}
```

## Best practices [#best-practices]

* **Store entity GUIDs:** Save the `entityGuid` returned from create operations. You'll need it for reading, updating, and deleting notebooks.
* **Validate JSON before upload:** Ensure your notebook payload is valid JSON and conforms to the `version` schema before sending.

> * **Use descriptive names:** Notebook names must be unique within an organization, so choose names that clearly indicate purpose (for example, `prod-checkout-investigation` rather than `notebook-1`).

Reviewer: ```suggestion
* **Use descriptive names:** <DNT>Notebooks</DNT> names must be unique within an organization, so choose names that clearly indicate purpose (for example, `prod-checkout-investigation` rather than `notebook-1`).
```
